### DISCUSSION TITLE
Program Not Printing To Screen

### DISCUSSION THREAD

**Discussion by mkbahr** [robhagemans/pcbasic#238] [General]
Greetings. I'm a rookie at programming, so please bear with my inexperience. 

I have a program that I think was written in GW-BASIC, and I want to verify that it works before attempting to translate it to a different language, as part of a research effort. The program is supposed to print to the screen using the PRINT command. It seemingly works, as I get no error messages upon running it and its one INPUT statement works, but nothing prints to the screen except "Ok". 

I would like to know if there's some setting I have to toggle that allows printing to the screen while running.
I also have no idea how to get a program to write to a file as well as print to the screen. 

The program is attached (zipped, because I guess GitHub doesn't allow for the uploading of .bas files), as is a "working" input file.
[738CODE.zip](https://github.com/robhagemans/pcbasic/files/15181044/738CODE.zip)
[inputvalue3.txt](https://github.com/robhagemans/pcbasic/files/15181063/inputvalue3.txt)

Thank you.

**Comment by robhagemans** [comment]
There's no setting needed to print to the screen. 

This program is indeed written in GW-BASIC. I tried the program with the given inputs both in PC-BASIC and in GW-BASIC and there is no output. Tracing the execution with TRON shows that with these inputs it doesn't actually reach any of the program lines that try to print anything to the screen - it just performs some calculations and exits.

Perhaps it was designed for the user to simply access the calculation results directly afterwards by printing the desired variable - or the program has been modified by someone to do so? E.g. the `END` on line 5395 might have been added later. But removing it just leads to more errors, at least with these inputs. There does seem to be at least one syntax error (`THEM` on line 1650) but it is not reached with the current input.

Your best bet may be to try to find another source for the same program and compare differences. This seems to be a variant of the RATEIEEE program distributed with IEEE 738-1986 which seems to be paywalled - a web search turns up several other variants including
- a 1997 version, but it needs QuickBASIC https://16557801101891546621.googlegroups.com/attach/33f54b7b2d3e9/IEEE%20738-2006%20%20_Bare%20Overhead%20Conductors_.pdf?part=0.6&vt=ANaJVrHeemhBCOd6d1aYaYrZRraNfSUmWsh62t2coLn6Tnt-AEV9bvNtrl7xGaC3fvB9sgqT-lHLm7Rq7L8YvfDjrsvkbbkoCSrt4ZQ_04D9YC2Qtx5tyQY
- another 1993 version in what appears to be GW-BASIC, which appears different to the one you shared http://normas.gers.co/ieee/ieee_std.738-1993.pdf

**Reply by mkbahr** [reply]
Thank you for the reply!

You are correct that this code is from IEEE Std. 738. It's a copy-paste from the 2012 revision of the standard, with a few empty line numbers removed. Like was said in the Marrin's reply to my post (confirming the slight suspicions I had), the code is likely incomplete. I imagine that is why it was omitted from the 2023 version. 

I'll take a look at the GW-BASIC version of the 1993 code and see how it works.

Again, thank you.

Just in case, on the off chance that PC-BASIC is used as part of a paper, how would you like to be credited?

**Reply by robhagemans** [reply]
Hi, I found a (dodgy) copy of the 2012 standard online and the program in there is slightly different in a few places, so there may have been some OCR issues or similar in copying the listing from PDF.

For example line 5495 is END IF, not END. That suggests that this is jn fact a QBASIC or QuickBASIC program - probably adjusted from an earlier GW-BASIC version, hence the line numbers. GW-BASIC (and PC-BASIC) do not have multiline IF / END IF blocks.

So you likely won't be able to use PC-BASIC even after the OCR errors are fixed - you could try FreeBASIC or QB64 which I think are QuickBASIC compatibles.

If you do want to cite PC-BASIC in a paper, a standard citation for software would make most sense, something along these lines https://blog.apastyle.org/apastyle/2015/01/how-to-cite-software-in-apa-style.html

**Reply by Marrin** [reply]
It's not QBASIC. QBASIC has `END IF` but to use that the code for `THEN` and `ELSE` must start on a new line and not directly after those keywords.  For instance this has still the problem that line 5380 leads to an error „`ELSE` without `IF`“:

```VB
REM * CALCULATE SOLAR AZIMUTH CONSTANT, CAZ
IF HOUR.ANG.DEG < 0 AND CHI < 0 THEN CAZ = 0
ELSE IF HOUR.ANG.DEG >= 0 AND CHI < 0 THEN CAZ = 360
ELSE CAS = 180
END IF
```

It should be written as either:

```VB
REM * CALCULATE SOLAR AZIMUTH CONSTANT, CAZ
IF HOUR.ANG.DEG < 0 AND CHI < 0 THEN
  CAZ = 0
ELSE
  IF HOUR.ANG.DEG >= 0 AND CHI < 0 THEN
    CAZ = 360
  ELSE
    CAZ = 180
  END IF
END IF
```

Or as:

```VB
REM * CALCULATE SOLAR AZIMUTH CONSTANT, CAZ
IF HOUR.ANG.DEG < 0 AND CHI < 0 THEN
  CAZ = 0
ELSE
  IF HOUR.ANG.DEG >= 0 AND CHI < 0 THEN CAZ = 360 ELSE CAZ = 180
END IF
```

**Comment by Marrin** [comment]
(This comment contains information already given by @robhagemans because I didn't came around to actually send it yesterday)

When started the code ends in line 5395 at an `END` statement in the middle of a subroutine.  It looks like someone put that there while developing/changing the program, maybe for debugging purposes.  Because the lines directly before that `END` are buggy.  `ELSE` is not supposed to be the very first statement in a line.  If there is no preceding `IF` it acts more or less as `REM`, i.e. the rest of the line is simply ignored.

It's also a bit odd that the code sets `CAZ` in two ”branches” but `CAS` in the third ”branch”.  Quotes because that is assuming the syntax would be correct.

When continuing the program after the `END` with the `CONT` command, it runs a bit further until it stops in line 15181 with a „Bad file number“ error.  The program tries to write some values to a file with file number #2, but that file is never opened anywhere in the code.  Also the `DEBUG` variable that is tested before trying to write, isn't assigned anywhere.

The program seems to be incomplete or in a state where someone tried to change or extend it, but did not finish that task.

The code contains the variable `NSELECT` which is used in several conditions but is never set to a value.  The data file seems to start with a value that is ”commented” as NSELECT, but is read into another variable.  It seems that data file doesn't belong to this code, or the code was altered but the data file wasn't changed accordingly.
